### PR TITLE
Refactor ivy.binary_cross_entropy 

### PR DESCRIPTION
**Pull request by akshatvishu**
At its present state the `ivy.binary_cross_entropy ()` function was missing few parameters which were available in it's `tf.keras` or `torch` counter-part hence, as part of the discussion at [discord](https://discord.com/channels/799879767196958751/998782045494976522/1107969223642329138) , I added additional parameters to ivy's `binary_cross_entropy ` function!

All test-case are passing! 

Here is the new function signature of `ivy.binary_cross_entropy ()`:
```python
def binary_cross_entropy(
    true: Union[ivy.Array, ivy.NativeArray],
    pred: Union[ivy.Array, ivy.NativeArray],
    /,
    *,
    from_logits: bool = False,
    epsilon: float = 0.0,  
    reduction: str = "none",
    pos_weight: Optional[Union[ivy.Array, ivy.NativeArray]] = None,
    axis: Optional[int] = None,
    out: Optional[ivy.Array] = None,
) -> ivy.Array:
```

 Old function signature:
 ```python
 def binary_cross_entropy(
    true: Union[ivy.Array, ivy.NativeArray],
    pred: Union[ivy.Array, ivy.NativeArray],
    /,
    *,
    epsilon: float = 1e-7,
    reduction: str = "none",
    out: Optional[ivy.Array] = None,
) -> ivy.Array:
 ```

### CODE REVIEW

**Comment by AnnaTz** [comment]
> @AnnaTz looping you in on this PR since you've been discussing this with the contributor. Do we want to extend the functionalities of `binary_cross_entropy` despite `sparse_cross_entropy` and `binary_cross_entropy_with_logits` already covering these cases ?

Hi @nassimberrada, I think this PR should be 1) unifying `binary_cross_entropy` and `binary_cross_entropy_with_logits`, and 2) extending with the `axis` and `pos_weights` arguments. I don't see any overlap with `sparse_cross_entropy` though.
@akshatvishu I think you should be removing `binary_cross_entropy_with_logits` in this PR as well to avoid confusion.

File: ivy_tests/test_ivy/test_functional/test_nn/test_losses.py [hunk 1]
```diff
@@ -70,54 +70,96 @@ def test_cross_entropy(
...
+        exclude_min=True,
+        exclude_max=True,
+        min_num_dims=1,
+        max_num_dims=1,
+        min_dim_size=2,
+        shape=(5,),
     ),
     reduction=st.sampled_from(["none", "sum", "mean"]),
+    axis=helpers.ints(min_value=-1, max_value=0),
     epsilon=helpers.floats(min_value=0, max_value=0.49),
```
[the remark this answers is not in the document]
Comment (akshatvishu) [reply ivy_tests/test_ivy/test_functional/test_nn/test_losses.py:107]: I will update it to check the full-range of valid values!

File: ivy/functional/ivy/losses.py [hunk 2]
```diff
@@ -183,13 +198,51 @@ def binary_cross_entropy(
...
-        out,
-    )
+
+    if not (0.0 <= epsilon <= 1.0):
+        raise ValueError("epsilon should be a float in [0, 1]")
+
+    if not from_logits and pos_weight is not None:
+        raise ValueError("pos_weight is only allowed when from_logits is set to True")
+
+    true = true.astype(pred.dtype)
```
[the remark this answers is not in the document]
Comment (akshatvishu) [reply ivy/functional/ivy/losses.py:208]: Thanks for pointing it out,  I will remove it!

File: ivy/functional/ivy/losses.py [hunk 3]
```diff
@@ -183,13 +198,51 @@ def binary_cross_entropy(
...
+
+    if from_logits:
+        if pos_weight is not None:
+            num_classes = pred.shape[0] if len(pred.shape) == 1 else pred.shape[1]
+            if pos_weight.shape[0] != num_classes:
+                raise ValueError(
+                    "pos_weight must have the same size as the number of classes in"
+                    " pred at non-singleton dimension 1"
+                )
+            epsilon_ = 1e-7
```
[the remark this answers is not in the document]
Comment (akshatvishu) [reply ivy/functional/ivy/losses.py:223]: Actually, the current version of `ivy.binary_cross_entropy()` used `epsilon` as an alias for `label-smoothing`.  Thus, I let it be called `epsilon` instead of changing it to `label-smoothing`.

The `epsilon_` is different from epsilon  and  is taken from TensorFlow implementation of the [same function](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L5035) and it values is 
```python
# Epsilon fuzz factor used throughout the codebase.
_EPSILON = 1e-7
```
defined at `/opt/fw/tensorflow/keras/backend_config.py`  and they've added this small value of ` epsilon_`  to the `output` and `1-output` when computing the logarithm to avoid taking the logarithm of `0`. This is because the logarithm of 0 is undefined and would result in an `error`. Adding a small value of epsilon ensures that the argument of the logarithm is always greater than 0, avoiding this issue.

Thus, the whole  source of confusion is because  we are using epsilon as an alias for label-smoothing!

**Comment by akshatvishu** [comment]
> Regarding the points you mentioned above, you only need to concern yourself with failures related to your implementation. Among those only (25) does. With that said, could you adjust the tests such that pred and true have the same dtypes when generating examples ?

@nassimberrada  Should I open a new PR for it !

> Finally, could you add some examples to the docstrings with variations of the variables you added ? Thanks !

Sure!

**Comment by akshatvishu** [comment]
@nassimberrada Added all the requested changes! Let me know if there is anything more to work upon!

**Comment by akshatvishu** [comment]
> LGTM ! Could you also remove the testing for `binary_cross_entropy_with_logits` in `test_experimental_/test_nn/test_losses` since you removed the function ?

Sure!

> A final detail, I think you've updated the torch frontend function for `binary_cross_entropy_with_logits` right ? Did you also update the `binary_cross_entropy` frontend function above it ?

No!

> If not could you update it as well if it needs to ? 
Sure!
>Otherwise let me know when you've removed the testing for the with_logits function, I can then merge your PR. Thanks !
just let me finish the above task quickly !

**Comment by akshatvishu** [comment]
> Looks good to be shipped, merging. Again, great work and thank you for your patience !

Thanks for the extra-fast  reviews ! 


image

^This is how I feel to see this getting merged
